client/session.go

```diff
@@ -563,3 +595,936 @@ func (s *session) Close() error {
+		}
+
+		// Determine maximum size
+		currMaxSize := int64(math.MinInt64)
+		for i := 0; i < len(currEligible); i++ {
+			unselected := currEligible[i].unselectedBlocks()
+			// Track maximum size
+			if unselected[0].size > currMaxSize {
+				currMaxSize = unselected[0].size
+			}
+		}
+
+		// Only select from those with the maximum size
+		for i := len(currEligible) - 1; i >= 0; i-- {
+			unselected := currEligible[i].unselectedBlocks()
+			if unselected[0].size < currMaxSize {
+				// Swap current entry to tail
+				blocksMetadatas(currEligible).swap(i, len(currEligible)-1)
+				// Trim newly last entry
+				currEligible = currEligible[:len(currEligible)-1]
+			}
+		}
+
+		// Order by least outstanding blocks being fetched
+		blocksMetadataQueues = blocksMetadataQueues[:0]
+		for i := range currEligible {
+			insert := blocksMetadataQueue{
+				blocksMetadata: currEligible[i],
+				queue:          peerQueues.findQueue(currEligible[i].peer),
+			}
+			blocksMetadataQueues = append(blocksMetadataQueues, insert)
+		}
+		sort.Stable(blocksMetadatasQueuesByOutstanding(blocksMetadataQueues))
+
+		// Select the best peer
+		bestPeerBlocksQueue := blocksMetadataQueues[0].queue
+		blocksMetadataQueues = blocksMetadataQueues[:0]
+
+		// Prepare the reattempt peers metadata
+		peersMetadata := make([]blockMetadataReattemptPeerMetadata, 0, len(currStart))
+		for i := range currStart {
+			unselected := currStart[i].unselectedBlocks()
+			metadata := blockMetadataReattemptPeerMetadata{
+				peer:  currStart[i].peer,
+				start: unselected[0].start,
+				size:  unselected[0].size,
+			}
+			peersMetadata = append(peersMetadata, metadata)
+		}
+
+		// Remove the block from all other peers and increment index for selected peer
+		for i := range currStart {
+			peer := currStart[i].peer
+			if peer == bestPeerBlocksQueue.peer {
+				// Select this block
+				idx := currStart[i].idx
+				currStart[i].idx = idx + 1
+
+				// Set the reattempt metadata
+				time.Sleep(time.Millisecond)
+				currStart[i].blocks[idx].reattempt.attempt++
+				currStart[i].blocks[idx].reattempt.attempted =
+					append(currStart[i].blocks[idx].reattempt.attempted, peer)
+				currStart[i].blocks[idx].reattempt.peersMetadata = peersMetadata
+
+				// Leave the block in the current peers blocks list
+				continue
+			}
+
+			// Removing this block
+			blocksLen := len(currStart[i].blocks)
+			idx := currStart[i].idx
+			tailIdx := blocksLen - 1
+			currStart[i].blocks[idx], currStart[i].blocks[tailIdx] = currStart[i].blocks[tailIdx], currStart[i].blocks[idx]
+			currStart[i].blocks = currStart[i].blocks[:blocksLen-1]
+		}
+	}
+
+	return nil
+}
+
+func (s *session) streamBlocksBatchFromPeer(
+	retrier xretry.Retrier,
+	peer hostQueue,
+	batch []*blocksMetadata,
+	blocksResult *blocksResult,
+	enqueueCh *enqueueChannel,
+) {
+	// Prepare request
+	var (
+		req    = rpc.NewFetchBlocksRequest()
+		result *rpc.FetchBlocksResult_
+	)
+	for i := range batch {
+		var starts []int64
```

Comment:
Nit: make this with len(batch[i].blocks) and then just assign to position j in following loop?